File: tests/algorithms/x/Java/machine_learning/decision_tree.java
public class Main {
    interface Tree {}

    static class Leaf implements Tree {
        double prediction;
        Leaf(double prediction) {
            this.prediction = prediction;
        }
        Leaf() {}
        @Override public String toString() {
            return String.format("{'prediction': %s}", String.valueOf(prediction));
        }
    }

    static class Branch implements Tree {
        double decision_boundary;
        Tree left;
        Tree right;
        Branch(double decision_boundary, Tree left, Tree right) {
            this.decision_boundary = decision_boundary;
            this.left = left;
            this.right = right;
        }
        Branch() {}
        @Override public String toString() {
            return String.format("{'decision_boundary': %s, 'left': %s, 'right': %s}", String.valueOf(decision_boundary), String.valueOf(left), String.valueOf(right));
        }
    }

    static double PI = (double)(3.141592653589793);
    static double TWO_PI = (double)(6.283185307179586);
    static long seed = 123456789L;

    static double _mod(double x, double m) {
        return (double)(x) - (double)((double)((((Number)(((Number)((double)(x) / (double)(m))).intValue())).doubleValue())) * (double)(m));
    }

    static double sin(double x) {
        double y = (double)((double)(_mod((double)((double)(x) + (double)(PI)), (double)(TWO_PI))) - (double)(PI));
        double y2_1 = (double)((double)(y) * (double)(y));
        double y3_1 = (double)((double)(y2_1) * (double)(y));
        double y5_1 = (double)((double)(y3_1) * (double)(y2_1));
        double y7_1 = (double)((double)(y5_1) * (double)(y2_1));
        return (double)((double)((double)(y) - (double)((double)(y3_1) / (double)(6.0))) + (double)((double)(y5_1) / (double)(120.0))) - (double)((double)(y7_1) / (double)(5040.0));
    }

    static double rand() {
        seed = (long)(((long)(Math.floorMod(((long)(((long)(1103515245L * (long)(seed)) + 12345L))), 2147483648L))));
        return (double)(((Number)(seed)).doubleValue()) / (double)(2147483648.0);
    }

    static double mean(double[] vals) {
        double sum = (double)(0.0);
        long i_1 = 0L;
        while ((long)(i_1) < (long)(vals.length)) {
            sum = (double)((double)(sum) + (double)(vals[(int)((long)(i_1))]));
            i_1 = (long)((long)(i_1) + 1L);
        }
        return (double)(sum) / (double)(vals.length);
    }

    static double mean_squared_error(double[] labels, double prediction) {
        double total = (double)(0.0);
        long i_3 = 0L;
        while ((long)(i_3) < (long)(labels.length)) {
            double diff_1 = (double)((double)(labels[(int)((long)(i_3))]) - (double)(prediction));
            total = (double)((double)(total) + (double)((double)(diff_1) * (double)(diff_1)));
            i_3 = (long)((long)(i_3) + 1L);
        }
        return (double)(total) / (double)(labels.length);
    }

    static Tree train_tree(double[] x, double[] y, long depth, long min_leaf_size) {
        if ((long)(x.length) < (long)(2L * (long)(min_leaf_size))) {
            return ((Tree)(new Leaf(mean(((double[])(y))))));
        }
        if ((long)(depth) == 1L) {
            return ((Tree)(new Leaf(mean(((double[])(y))))));
        }
        long best_split_1 = 0L;
        double min_error_1 = (double)((double)(mean_squared_error(((double[])(x)), (double)(mean(((double[])(y)))))) * (double)(2.0));
        long i_5 = 0L;
        while ((long)(i_5) < (long)(x.length)) {
            if ((long)(java.util.Arrays.copyOfRange(x, (int)(0L), (int)((long)(i_5))).length) < (long)(min_leaf_size)) {
                i_5 = (long)(i_5);
            } else             if ((long)(java.util.Arrays.copyOfRange(x, (int)((long)(i_5)), (int)((long)(x.length))).length) < (long)(min_leaf_size)) {
                i_5 = (long)(i_5);
            } else {
                double err_left_1 = (double)(mean_squared_error(((double[])(java.util.Arrays.copyOfRange(x, (int)(0L), (int)((long)(i_5))))), (double)(mean(((double[])(java.util.Arrays.copyOfRange(y, (int)(0L), (int)((long)(i_5)))))))));
                double err_right_1 = (double)(mean_squared_error(((double[])(java.util.Arrays.copyOfRange(x, (int)((long)(i_5)), (int)((long)(x.length))))), (double)(mean(((double[])(java.util.Arrays.copyOfRange(y, (int)((long)(i_5)), (int)((long)(y.length)))))))));
                double err_1 = (double)((double)(err_left_1) + (double)(err_right_1));
                if ((double)(err_1) < (double)(min_error_1)) {
                    best_split_1 = (long)(i_5);
                    min_error_1 = (double)(err_1);
                }
            }
            i_5 = (long)((long)(i_5) + 1L);
        }
        if ((long)(best_split_1) != 0L) {
            double[] left_x_1 = ((double[])(java.util.Arrays.copyOfRange(x, (int)(0L), (int)((long)(best_split_1)))));
            double[] left_y_1 = ((double[])(java.util.Arrays.copyOfRange(y, (int)(0L), (int)((long)(best_split_1)))));
            double[] right_x_1 = ((double[])(java.util.Arrays.copyOfRange(x, (int)((long)(best_split_1)), (int)((long)(x.length)))));
            double[] right_y_1 = ((double[])(java.util.Arrays.copyOfRange(y, (int)((long)(best_split_1)), (int)((long)(y.length)))));
            double boundary_1 = (double)(x[(int)((long)(best_split_1))]);
            Tree left_tree_1 = train_tree(((double[])(left_x_1)), ((double[])(left_y_1)), (long)((long)(depth) - 1L), (long)(min_leaf_size));
            Tree right_tree_1 = train_tree(((double[])(right_x_1)), ((double[])(right_y_1)), (long)((long)(depth) - 1L), (long)(min_leaf_size));
            return ((Tree)(new Branch(boundary_1, left_tree_1, right_tree_1)));
        }
        return ((Tree)(new Leaf(mean(((double[])(y))))));
    }

    static double predict(Tree tree, double value) {
        return tree instanceof Leaf ? ((Leaf)(tree)).prediction : (double)(value) >= ((Number)(((Branch)(tree)).decision_boundary)).doubleValue() ? predict(((Branch)(tree)).right, (double)(value)) : predict(((Branch)(tree)).left, (double)(value));
    }

    static void main() {
        double[] x = ((double[])(new double[]{}));
        double v_1 = (double)(-1.0);
        while ((double)(v_1) < (double)(1.0)) {
            x = ((double[])(appendDouble(x, (double)(v_1))));
            v_1 = (double)((double)(v_1) + (double)(0.005));
        }
        double[] y_2 = ((double[])(new double[]{}));
        long i_7 = 0L;
        while ((long)(i_7) < (long)(x.length)) {
            y_2 = ((double[])(appendDouble(y_2, (double)(sin((double)(x[(int)((long)(i_7))]))))));
            i_7 = (long)((long)(i_7) + 1L);
        }
        Tree tree_1 = train_tree(((double[])(x)), ((double[])(y_2)), 10L, 10L);
        double[] test_cases_1 = ((double[])(new double[]{}));
        i_7 = 0L;
        while ((long)(i_7) < 10L) {
            test_cases_1 = ((double[])(appendDouble(test_cases_1, (double)((double)((double)(rand()) * (double)(2.0)) - (double)(1.0)))));
            i_7 = (long)((long)(i_7) + 1L);
        }
        double[] predictions_1 = ((double[])(new double[]{}));
        i_7 = 0L;
        while ((long)(i_7) < (long)(test_cases_1.length)) {
            predictions_1 = ((double[])(appendDouble(predictions_1, (double)(predict(tree_1, (double)(test_cases_1[(int)((long)(i_7))]))))));
            i_7 = (long)((long)(i_7) + 1L);
        }
        double sum_err_1 = (double)(0.0);
        i_7 = 0L;
        while ((long)(i_7) < (long)(test_cases_1.length)) {
            double diff_3 = (double)((double)(predictions_1[(int)((long)(i_7))]) - (double)(test_cases_1[(int)((long)(i_7))]));
            sum_err_1 = (double)((double)(sum_err_1) + (double)((double)(diff_3) * (double)(diff_3)));
            i_7 = (long)((long)(i_7) + 1L);
        }
        double avg_error_1 = (double)((double)(sum_err_1) / (double)(test_cases_1.length));
        System.out.println("Test values: " + _p(test_cases_1));
        System.out.println("Predictions: " + _p(predictions_1));
        System.out.println("Average error: " + _p(avg_error_1));
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static double[] appendDouble(double[] arr, double v) {
        double[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/PHP/machine_learning/decision_tree.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  $PI = 3.141592653589793;
  $TWO_PI = 6.283185307179586;
  function _mod($x, $m) {
  global $PI, $TWO_PI, $seed;
  return $x - (floatval(intval($x / $m))) * $m;
};
  function mochi_sin($x) {
  global $PI, $TWO_PI, $seed;
  $y = _mod($x + $PI, $TWO_PI) - $PI;
  $y2 = $y * $y;
  $y3 = $y2 * $y;
  $y5 = $y3 * $y2;
  $y7 = $y5 * $y2;
  return $y - $y3 / 6.0 + $y5 / 120.0 - $y7 / 5040.0;
};
  $seed = 123456789;
  function mochi_rand() {
  global $PI, $TWO_PI, $seed;
  $seed = (1103515245 * $seed + 12345) % 2147483648;
  return floatval($seed) / 2147483648.0;
};
  function mean($vals) {
  global $PI, $TWO_PI, $seed;
  $sum = 0.0;
  $i = 0;
  while ($i < count($vals)) {
  $sum = $sum + $vals[$i];
  $i = $i + 1;
};
  return $sum / count($vals);
};
  function mean_squared_error($labels, $prediction) {
  global $PI, $TWO_PI, $seed;
  $total = 0.0;
  $i = 0;
  while ($i < count($labels)) {
  $diff = $labels[$i] - $prediction;
  $total = $total + $diff * $diff;
  $i = $i + 1;
};
  return $total / count($labels);
};
  function train_tree($x, $y, $depth, $min_leaf_size) {
  global $PI, $TWO_PI, $seed;
  if (count($x) < 2 * $min_leaf_size) {
  return ['__tag' => 'Leaf', 'prediction' => mean($y)];
}
  if ($depth == 1) {
  return ['__tag' => 'Leaf', 'prediction' => mean($y)];
}
  $best_split = 0;
  $min_error = mean_squared_error($x, mean($y)) * 2.0;
  $i = 0;
  while ($i < count($x)) {
  if (count(array_slice($x, 0, $i)) < $min_leaf_size) {
  $i = $i;
} else {
  if (count(array_slice($x, $i)) < $min_leaf_size) {
  $i = $i;
} else {
  $err_left = mean_squared_error(array_slice($x, 0, $i), mean(array_slice($y, 0, $i)));
  $err_right = mean_squared_error(array_slice($x, $i), mean(array_slice($y, $i)));
  $err = $err_left + $err_right;
  if ($err < $min_error) {
  $best_split = $i;
  $min_error = $err;
};
};
}
  $i = $i + 1;
};
  if ($best_split != 0) {
  $left_x = array_slice($x, 0, $best_split);
  $left_y = array_slice($y, 0, $best_split);
  $right_x = array_slice($x, $best_split);
  $right_y = array_slice($y, $best_split);
  $boundary = $x[$best_split];
  $left_tree = train_tree($left_x, $left_y, $depth - 1, $min_leaf_size);
  $right_tree = train_tree($right_x, $right_y, $depth - 1, $min_leaf_size);
  return ['__tag' => 'Branch', 'decision_boundary' => $boundary, 'left' => $left_tree, 'right' => $right_tree];
}
  return ['__tag' => 'Leaf', 'prediction' => mean($y)];
};
  function predict($tree, $value) {
  global $PI, $TWO_PI, $seed;
  return (function($__v) {
  if ($__v['__tag'] === "Leaf") {
    $p = $__v["prediction"];
    return $p;
  } elseif ($__v['__tag'] === "Branch") {
    $b = $__v["decision_boundary"];
    $l = $__v["left"];
    $r = $__v["right"];
    return ($value >= $b ? predict($r, $value) : predict($l, $value));
  }
})($tree);
};
  function main() {
  global $PI, $TWO_PI, $seed;
  $x = [];
  $v = -1.0;
  while ($v < 1.0) {
  $x = _append($x, $v);
  $v = $v + 0.005;
};
  $y = [];
  $i = 0;
  while ($i < count($x)) {
  $y = _append($y, mochi_sin($x[$i]));
  $i = $i + 1;
};
  $tree = train_tree($x, $y, 10, 10);
  $test_cases = [];
  $i = 0;
  while ($i < 10) {
  $test_cases = _append($test_cases, mochi_rand() * 2.0 - 1.0);
  $i = $i + 1;
};
  $predictions = [];
  $i = 0;
  while ($i < count($test_cases)) {
  $predictions = _append($predictions, predict($tree, $test_cases[$i]));
  $i = $i + 1;
};
  $sum_err = 0.0;
  $i = 0;
  while ($i < count($test_cases)) {
  $diff = $predictions[$i] - $test_cases[$i];
  $sum_err = $sum_err + $diff * $diff;
  $i = $i + 1;
};
  $avg_error = $sum_err / count($test_cases);
  echo rtrim('Test values: ' . _str($test_cases)), PHP_EOL;
  echo rtrim('Predictions: ' . _str($predictions)), PHP_EOL;
  echo rtrim('Average error: ' . _str($avg_error)), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
